### PR TITLE
expression: throw "too big precision" error for CAST(AS TIME)

### PR DESCRIPTION
**Pull request by zz-jason**
### What problem does this PR solve? 

For the following SQL:
```sql
SELECT CAST(1 AS TIME(31));
```
we should return a error as MySQL does:
```sql
MySQL(localhost:3306) > SELECT CAST(1 AS TIME(31));
ERROR 1426 (42000): Too-big precision 31 specified for 'CAST'. Maximum is 6.
```

Another tiny bugfix is, for cast decimal as time, when the input decimal is NULL, we should return NULL.

### What is changed and how it works?

Check the target precision first, if the check result is positive, we then continue to recursively evaluate child expression.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8907)

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -999,9 +1005,12 @@ func (b *builtinCastDecimalAsDurationSig) Clone() builtinFunc {
 }
 
 func (b *builtinCastDecimalAsDurationSig) evalDuration(row chunk.Row) (res types.Duration, isNull bool, err error) {
+	if b.tp.Decimal > types.MaxFsp {
+		return res, true, errTooBigPrecision.GenWithStackByArgs(b.tp.Decimal, "CAST", types.MaxFsp)
+	}
 	val, isNull, err := b.args[0].EvalDecimal(b.ctx, row)
 	if isNull || err != nil {
-		return res, false, err
+		return res, true, err
```
Comment (winoros) [inline expression/builtin_cast.go]:
```suggestion
		return res, isNull, err
```
is better?

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -570,6 +570,9 @@ func (b *builtinCastIntAsDurationSig) Clone() builtinFunc {
 }
 
 func (b *builtinCastIntAsDurationSig) evalDuration(row chunk.Row) (res types.Duration, isNull bool, err error) {
+	if b.tp.Decimal > types.MaxFsp {
```
Comment (eurekaka) [inline expression/builtin_cast.go]: Can we put this check in `Preprocess`? that may look cleaner. Seems decimal of `time` here must be a constant, so it is safe to do check in `Preprocess`? here is the output of MySQL:
```
mysql> select * from s;
+------+------+
| a    | b    |
+------+------+
|    1 | NULL |
|    2 |    1 |
+------+------+
2 rows in set (0.00 sec)

mysql> select cast(1 as time(a)) from s;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'a)) from s' at line 1
```

File: expression/builtin_cast.go (hunk 1 above)
Comment (zz-jason) [reply expression/builtin_cast.go]: I think directly use `true` is better for understanding.

File: expression/builtin_cast.go (hunk 2 above)
Comment (zz-jason) [reply expression/builtin_cast.go]: Let me have a try.

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -570,6 +570,9 @@ func (b *builtinCastIntAsDurationSig) Clone() builtinFunc {
...
+		return res, true, errTooBigPrecision.GenWithStackByArgs(b.tp.Decimal, "CAST", types.MaxFsp)
```
Comment (qw4990) [inline expression/builtin_cast.go]: isNull should be false on this error case?

```suggestion
		return res, false, errTooBigPrecision.GenWithStackByArgs(b.tp.Decimal, "CAST", types.MaxFsp)
```
is better?

File: planner/core/expression_rewriter.go [hunk 4]
```diff
@@ -793,10 +793,18 @@ func (er *expressionRewriter) Leave(originInNode ast.Node) (retNode ast.Node, ok
 		er.caseToExpression(v)
 	case *ast.FuncCastExpr:
 		arg := er.ctxStack[len(er.ctxStack)-1]
+
```
Comment (XuHuaiyu) [inline planner/core/expression_rewriter.go]: useless line

Comment (zz-jason) [reply planner/core/expression_rewriter.go]: adressed, PTAL

**Pull request merged**